### PR TITLE
Metrics from Health handlers for Bots V2

### PR DESCRIPTION
**Pull request by dkeysil**
Since bots v2 are sending alerts by themselves but no metrics - we need to introduce new way of sending metrics - by polling health handler to gather additional information.

These metrics then are going to be aggregated and indexed.

Changed:
- Each metric now can contain a chain id (and one V2 bot can expose metrics related to many chains in the same time)
- It's possible to pass metrics in addition to errors in the health handler

### CODE REVIEW

**Comment by aomerk** [comment]
i think it would be useful if we had an e2e test for this one:
- choose one of the bots under `tests/e2e/agents`
- implement health check
- add a test case for health checks under `e2e_forta_local_mode_test.go`

File: services/components/botio/bot_client.go [hunk 1]
```diff
@@ -545,7 +545,8 @@ func (bot *botClient) processTransaction(ctx context.Context, lg *log.Entry, req
 		// truncate findings
 		if len(resp.Findings) > MaxFindings {
 			dropped := len(resp.Findings) - MaxFindings
-			droppedMetric := metrics.CreateAgentMetric(botConfig, metrics.MetricFindingsDropped, float64(dropped))
+			// TODO: for v1 bots include chain ID from the scanner config
```
Comment (canercidam) [inline services/components/botio/bot_client.go:548]: Is this done now? I see that `CreateAgentMetricV1()` seems to set it.

File: services/publisher/metrics.go [hunk 2]
```diff
@@ -39,15 +44,21 @@ func (mb *metricsBucket) CreateAndGetSummary(name string) *protocol.MetricSummar
...
 	}
 }
 
-func (ama *AgentMetricsAggregator) findBucket(agentID string, t time.Time) *metricsBucket {
+func (ama *AgentMetricsAggregator) findBucket(agentID string, chainID int64, t time.Time) *metricsBucket {
 	bucketTime := ama.FindClosestBucketTime(t)
-	for _, bucket := range ama.buckets {
+	buckets, ok := ama.bucketsByChainID[chainID]
+	if !ok {
+		ama.bucketsByChainID[chainID] = make([]*metricsBucket, 0)
```
Comment (canercidam) [inline services/publisher/metrics.go:58]:
```suggestion
		buckets = make([]*metricsBucket, 0)
		ama.bucketsByChainID[chainID] = buckets
```

Comment (canercidam) [reply services/publisher/metrics.go:58]: Does that make sense?

File: services/publisher/publisher.go [hunk 3]
```diff
@@ -155,6 +155,15 @@ func (pub *Publisher) publishNextBatch(batch *protocol.AlertBatch) (published bo
 		}
 	}
 
+	// add chain id if it's not set, e.g. EventMetric
+	for _, metric := range batch.Metrics {
```
[the remark this answers is not in the document]
Comment (canercidam) [reply services/publisher/publisher.go:159]: Also that would be great if we could set the chain ID in v2 bots' lifecycle metrics but only if it's not very difficult.

[the remark this answers is not in the document]
Comment (canercidam) [reply services/publisher/publisher.go:159]: The lifecycle metrics case is outside of what we want to support for the users and on the UI so we could consider those ones out of scope and optionally deal with v2 lifecycle metric chains IDs later.

File: services/publisher/publisher.go [hunk 4]
```diff
@@ -155,6 +155,16 @@ func (pub *Publisher) publishNextBatch(batch *protocol.AlertBatch) (published bo
 		}
 	}
 
+	// add chain id if it's not set, e.g. EventMetric
+	// TODO: based on the bot id query agent config and use chain id from it or do that inside the aggregator
+	for _, metric := range batch.Metrics {
```
Comment (canercidam) [inline services/publisher/publisher.go:160]: Shall we move this loop into the aggregator, inside `PrepareMetrics()`?

Comment (canercidam) [reply services/publisher/publisher.go:160]: The aggregator can be provided the scanner chain ID at initialization. (as discussed in the Huddle)

**Comment by aomerk** [comment]
Can we use metric name definitions from the core-go PR?

File: services/components/botio/bot_client_test.go [hunk 5]
```diff
@@ -96,15 +96,20 @@ func (s *BotClientSuite) TestStartProcessStop() {
...
-	s.lifecycleMetrics.EXPECT().HealthCheckSuccess(s.botClient.configUnsafe).MinTimes(1)
+	s.lifecycleMetrics.EXPECT().HealthCheckSuccess(s.botClient.configUnsafe).Do(func(_ interface{}) {
+		close(healthCheckChan)
+	}).MinTimes(1)
 
 	s.msgClient.EXPECT().Publish(messaging.SubjectAgentsAlertSubscribe, combinerSubscriptions)
 	s.botClient.StartProcessing()
 	s.botClient.Initialize()
 
+	<-healthCheckChan
```
Comment (canercidam) [inline services/components/botio/bot_client_test.go:111]: Thanks!

File: services/publisher/metrics.go [hunk 6]
```diff
@@ -75,20 +89,26 @@ func (ama *AgentMetricsAggregator) FindClosestBucketTime(t time.Time) time.Time
...
 
 	for _, m := range ms.Metrics {
 		t, _ := time.Parse(time.RFC3339, m.Timestamp)
-		bucket := ama.findBucket(m.AgentId, t)
-		bucket.MetricCounters[m.Name] = append(bucket.MetricCounters[m.Name], uint32(m.Value))
-		if m.Details != "" {
-			bucket.MetricDetails[m.Name] = m.Details
+
+		// add chain id if it's not set, e.g. EventMetric
+		chainID := ama.chainID
```
Comment (canercidam) [inline services/publisher/metrics.go:100]: Thanks for moving this here.

**Review by canercidam — APPROVED** [review]

**Pull request merged**